crates/assists/src/handlers/generate_enum_match_method.rs

```diff
@@ -34,79 +35,284 @@ use crate::{
 //     }
 // }
 // ```
-pub(crate) fn generate_enum_match_method(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+pub(crate) fn generate_enum_is_method(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
     let variant = ctx.find_node_at_offset::<ast::Variant>()?;
     let variant_name = variant.name()?;
-    let parent_enum = variant.parent_enum();
-    if !matches!(variant.kind(), ast::StructKind::Unit) {
-        mark::hit!(test_gen_enum_match_on_non_unit_variant_not_implemented);
-        return None;
-    }
+    let parent_enum = ast::Adt::Enum(variant.parent_enum());
+    let variant_kind = variant_kind(&variant);
 
     let enum_lowercase_name = to_lower_snake_case(&parent_enum.name()?.to_string());
-    let fn_name = to_lower_snake_case(&variant_name.to_string());
+    let fn_name = format!("is_{}", &to_lower_snake_case(variant_name.text()));
 
     // Return early if we've found an existing new fn
-    let impl_def = find_struct_impl(
-        &ctx,
-        &ast::Adt::Enum(parent_enum.clone()),
-        format!("is_{}", fn_name).as_str(),
-    )?;
+    let impl_def = find_struct_impl(&ctx, &parent_enum, &fn_name)?;
 
     let target = variant.syntax().text_range();
     acc.add(
-        AssistId("generate_enum_match_method", AssistKind::Generate),
+        AssistId("generate_enum_is_method", AssistKind::Generate),
         "Generate an `is_` method for an enum variant",
         target,
         |builder| {
-            let mut buf = String::with_capacity(512);
-
-            if impl_def.is_some() {
-                buf.push('\n');
-            }
-
             let vis = parent_enum.visibility().map_or(String::new(), |v| format!("{} ", v));
-            format_to!(
-                buf,
+            let method = format!(
                 "    /// Returns `true` if the {} is [`{}`].
-    {}fn is_{}(&self) -> bool {{
-        matches!(self, Self::{})
+    {}fn {}(&self) -> bool {{
+        matches!(self, Self::{}{})
     }}",
                 enum_lowercase_name,
                 variant_name,
                 vis,
                 fn_name,
-                variant_name
+                variant_name,
+                variant_kind.pattern_suffix(),
+            );
+
+            add_method_to_adt(builder, &parent_enum, impl_def, &method);
+        },
+    )
+}
+
+// Assist: generate_enum_into_method
+//
+// Generate an `into_` method for an enum variant.
+//
+// ```
+// enum Value {
+//  Number(i32),
+//  Text(String)$0,
+// }
+// ```
+// ->
+// ```
+// enum Value {
+//  Number(i32),
+//  Text(String),
+// }
+//
+// impl Value {
+//     fn into_text(self) -> Option<String> {
+//         if let Self::Text(v) = self {
+//             Some(v)
+//         } else {
+//             None
+//         }
+//     }
+// }
```

Comment:
## Function name

I would've expected the generated method to be parameterized by the name of the output type, rather than the name of the enum variant. `as` is always followed by the type we're converting into (e.g. [`slice::as_ptr`](https://doc.rust-lang.org/std/primitive.str.html#method.as_ptr) or [`ThreadId::as_u64`](https://doc.rust-lang.org/std/thread/struct.ThreadId.html#method.as_u64)).

```suggestion
// enum Value {
//  Number(i32),
//  Text(String)$0,
// }
// ```
// ->
// ```
// enum Value {
//  Number(i32),
//  Text(String),
// }
//
// impl Value {
//     fn into_string(self) -> Option<String> {
//         if let Self::Text(v) = self {
//             Some(v)
//         } else {
//             None
//         }
//     }
// }
```

## Output type

Running clippy on the generated function produces the following warning:

```
warning: methods called `into_*` usually take self by value; consider choosing a less ambiguous name
  --> src/lib.rs:15:20
   |
15 |     fn into_string(&self) -> Option<&String> {
   |                    ^^^^^
   |
   = note: `#[warn(clippy::wrong_self_convention)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention
```

This matches the [Rust naming guidelines](https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter) and means we should probably only ever generate this function if we know we will never fail the conversion.

Reply:
How did you get that? It does generate `into_*` methods with `self` as the parameter, and `as_*` methods with `&self`.

Reply:
I disagree with your point about the function names. Naming the function by the type is good as long as the function is about the type conversion. A key difference is that these functions are intended to extract the payload from a specific enum variant, so they are precisely about the *variant* instead of the *type*. And when the function is generated for a specific variant it is weird to name them by type when some other variants have the same payload type, or when the payload is a generic type parameter. Also I think that naming them by the type would also be inconsistent with `is_` functions which are very similar in purpose and are precise about variant.

The concrete example I had in mind when I saw the request for these was [rowan's `NodeOrToken` methods](https://docs.rs/rowan/0.12.5/rowan/enum.NodeOrToken.html#method.into_node). Another example is `Result::ok` and `Result::err` - they also use the name to indicate which variant is extracted, but don't use the `as_` and `into_` naming convention. (although I guess that the fact that both of these examples are generic enums does not really help my case as referring by type would be super weird)

So maybe the naming convention for these could be different - I don't know if there's actually a convention for enum payload extracting functions (by the way, the request in #7604 did as for `as_variant_name()`. But in any case I think that the function needs to name the variant name, and not the payload type.

Reply:
> (although I guess that the fact that both of these examples are generic enums does not really help my case as referring by type would be super weird)

Maybe we can extrapolate a rule which is consistent with both cases: if the type we're referencing is a generic value we should use the variant name. If it's not, we should reference the output type. I agree that `NodeOrToken::as_n()` and `Result::t()` wouldn't be very good names. However I also think that `Value::as_str` which returns `&str` is clearer than `Value::as_text`.

---

> It does generate into_* methods with self as the parameter

Ah apologies, that's my bad. I probably wrongly copied / edited in playground; we can indeed disregard my comments on clippy 😅 .

However I'm still somewhat uncomfortable with auto-generating `into` functions which return `Option`. This seems like a fallible conversion for which `try_into` would be a better name. But instead of returning `Option`, `Result` may be a better fit.

The `Result::Err` path here is important because it allow recovering the original type through the error path in case the conversion fails. An example of this is [`TryFrom<Vec<T, A>> for [T; N]`](https://doc.rust-lang.org/std/convert/trait.TryFrom.html#impl-TryFrom%3CVec%3CT%2C%20A%3E%3E) which has `Vec<T, A>` as the error type. In the `Rowan` example, if the conversion fails the original type is gone. The `Copy` bound helps work around that; but Rust-Analyzer shouldn't assume all structs to have `Copy` bounds, which makes this hard to generalize.

I'd prefer either `into_x` conversions which always succeed, or `try_into_x` conversions which return the original type in case they fail (or both?).

Reply:
I still disagree with naming by type. I looked for more examples, and I found a couple (who would have guessed that non-error enums would be so difficult to find):
* [`serde_json::Value`](https://docs.rs/serde_json/1.0.62/serde_json/enum.Value.html#method.is_number). This is a fun one. First of, it has `as_object`, `as_array`, and even `as_null`. These make much more sense than `as_map`, `as_vec`, or `as_unit` because they speak in json terminology, so when reading the conversions it is a lot more obvious about what json kind is expected. Naming by type could have also lead to some very non-useful names - `as_map` could have also ended up being `as_btree_map` or `as_index_map`, which is just silly for manipulating json values. Now the interesting part is that this enum also has `as_i64`, `as_u64`, and `as_f64`. These are clearly about type conversion - there is no variant which would be named like that or would have a payload of corresponding type. Instead these check both for `Number` variant, and also that the payload is actually representable as the requested type.
* I also grepped rust-lang/rust for similar pattern, and the only real example I found was [`SocketAddr::as_pathname`](https://doc.rust-lang.org/std/os/unix/net/struct.SocketAddr.html#method.as_pathname). Return type is `Path`, but the name says "pathname" because it actually checks if it a pathname address.

I think serde_json example also shows that conversion to a type can have a more complex meaning than just matching for a single variant and that meaning would not be appropriate for an assist to guess. Also, remember the case when multiple variants have the same payload type. It would be weird for a function named `as_str` to be only about one of the variants holding a string. How would you even name the function for the other variant?

I agree with your point about `try_into_x` though, being able to recover the original value is useful. I changed the assist to instead generate `try_into_x` returning `Result<Payload, Self>`.